### ISSUE TITLE
Suggest ?Sized on generics where applicable

### ISSUE THREAD

**Issue by J-F-Liu** [rust-lang/rust#38936]
I meet a compile error can be demonstrated by the following code, and think it's irrational:
```rust
trait Train {
    fn knots(&self) -> usize;
}

impl Train for str {
    fn knots(&self) -> usize {
        self.chars().count()
    }
}

fn train_length<T:Train>(text: &'static T) -> usize {
    text.knots()
}

fn str_length(text: &'static str) -> usize {
    text.knots()
}

fn main() {
    println!("{:?}", str_length("asdfas"));   // works
    println!("{:?}", train_length("asdfas")); // compile error
}
```

**Comment by kevinmehall** [comment]
Use `fn train_length<T:Train + ?Sized>(text: &'static T) -> usize`.

Bare `str` is an [unsized type](https://doc.rust-lang.org/stable/book/unsized-types.html) and generics assume that types are sized by default. Here's [another post explaining unsized types](http://huonw.github.io/blog/2015/01/the-sized-trait/).

**Comment by J-F-Liu** [comment]
Yes, add `+?Sized` can solve the problem, the compile message should be more helpful.

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> src/main.rs:23:35
   |
13 | fn train_length<T:Train>(text: &'static T) -> usize {
   |    ------------ - required by this bound in `train_length`
...
23 |     println!("{:?}", train_length("asdfas")); // compile error
   |                                   ^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
```

It should be 

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> src/main.rs:23:35
   |
13 | fn train_length<T:Train>(text: &'static T) -> usize {
   |    ------------ -      - help: relax the implicit restriction:  `+ ?Sized`
   |                 |
   |                 `Sized` implicitly required in this bound in `train_length`
...
23 |     println!("{:?}", train_length("asdfas")); // compile error
   |                                   ^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
```

**Comment by estebank** [comment]
Current output:

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
  --> file4.rs:21:35
   |
11 | fn train_length<T:Train>(text: &'static T) -> usize {
   |    ------------ -      - help: consider relaxing the implicit `Sized` restriction: `+  ?Sized`
   |                 |
   |                 required by this bound in `train_length`
...
21 |     println!("{:?}", train_length("asdfas")); // compile error
   |                                   ^^^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: to learn more, visit <https://doc.rust-lang.org/book/ch19-04-advanced-types.html#dynamically-sized-types-and-the-sized-trait>
```

Fixed in #68377.
